### PR TITLE
add copilot instructions

### PR DESCRIPTION
**Pull request by GarrettBeatty**
Add copilot custom instructions for copilot review. Customizing this file allows us to specifiy what copilot should look for during reviews

See: https://github.blog/changelog/2025-06-13-copilot-code-review-customization-for-all/ 

## Motivation and Context
To improve copilot reviews

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have read the **README** document
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

[issues]: https://github.com/aws/aws-sdk-net/issues
[license]: http://aws.amazon.com/apache2.0/
[cla]: http://en.wikipedia.org/wiki/Contributor_License_Agreement

### CODE REVIEW

File: .github/copilot-instructions.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+- Copilot should flag PRs that:
+ 	- Introduce breaking changes without a major version bump.
+ 	- Lack a dev config file or changelog update.
+ 	- Change public APIs without documentation or tests.
+ 	- Have failing unit or integration tests.
+ 	- Reduce test coverage for affected code.
+ 	- Introduce unhandled exceptions, missing error handling, or poor input validation.
+ 	- Increase code complexity or reduce maintainability.
+ 	- Do not document or justify changes in logic, edge cases, or optimizations.
+ 	- For generator changes, do not provide evidence of equivalence (e.g., missing test results, behavioral comparison, or documentation).
```
Comment (peterrsongg) [inline .github/copilot-instructions.md]: you meant to say "do not provide evidence" here?

Comment (GarrettBeatty) [reply .github/copilot-instructions.md]: yes. "copilot should flag changes that " ... "for generator changes, do not provide evidence of equivalence"

Comment (GarrettBeatty) [reply .github/copilot-instructions.md]: maybe to make it more clear, "for changes where we are moving custom code to generator". this was basically to try and capture more things in the s3 migration, but if thats basically complete now then maybe i can remove this?

Comment (peterrsongg) [reply .github/copilot-instructions.md]: it is not basically complete😅i just wanted to make sure you meant "do not" and not "do"

Comment (GarrettBeatty) [reply .github/copilot-instructions.md]: ill reword it to make it more clear. "For changes related to migrating custom code to generated code, where evidence of  equivalence (e.g., missing test results, behavioral comparison, or documentation) is not provided."

File: .github/copilot-instructions.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+### 1. API Compatibility
+- Ensure **no breaking changes** are introduced in public APIs.
+- All public methods, classes, and interfaces must remain backward compatible unless a major version bump is justified.
+
+### 2. Behavioral Changes
+- Check for **changes in behavior** (e.g., logic, side effects, error handling).
+- If behavior changes, ensure it is documented and justified in the PR description.
+
+### 3. Dev Config and Versioning
```
Comment (boblodgett) [inline .github/copilot-instructions.md:16]: I don't think it is even going to begin to understand what this is. You likely need to obtain some information about it since we don't have the memory-bank yet but agree it cannot be huge for this purpose.

https://github.com/aws/aws-sdk-net/blob/boblod-ai-memory-bank-v2/memory-bank/partitions/sdkdeveloper/devConfig.md

Comment (boblodgett) [reply .github/copilot-instructions.md:16]: Perhaps the location and a sample of a devconfig file would be enough to start with the possible names.

File: .github/copilot-instructions.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+### 4. Platform Compatibility
+- Ensure changes do not break support for the following .NET versions:
+	- .NET Standard 2.0
+	- .NET Core 3.1
+	- .NET 8.0
+	- .NET Framework 4.7.2
+
+### 5. General Best Practices
+- Follow .NET and AWS SDK coding standards.
```
Comment (boblodgett) [inline .github/copilot-instructions.md:32]: What is it going to do with this when we don't have a coding standards doc or even memory bank info for this yet? It may be better to tell it to maintain coding consistency at this point since there are no standard docs.

File: .github/copilot-instructions.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+- Avoid introducing unused code, dependencies, or files.
+- Ensure all code is properly formatted and linted.
+
+### 6. Generator Changes
```
Comment (boblodgett) [inline .github/copilot-instructions.md:36]: Is it going to understand you are referring to our generator (Doc or code)? I have a feeling it doesn't even know where that is at.

File: .github/copilot-instructions.md (hunk 2 above)
Comment (GarrettBeatty) [reply .github/copilot-instructions.md:16]: i have this other PR https://github.com/aws/aws-sdk-net/pull/3968 which has info on dev config and contributing. maybe i can referencing that CONTRIBUTING.MD file?

File: .github/copilot-instructions.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+- When replacing handwritten code with generated code, ensure the generated code is functionally equivalent to the original.
+- Document any differences, edge cases, or limitations in the PR description.
+- If possible, provide a comparison or summary of key behaviors between the old and new implementations.
+
+ ### 7. General Review Practices
```
Comment (boblodgett) [inline .github/copilot-instructions.md:41]: Add look for security vulnerabilities

File: .github/copilot-instructions.md (hunk 3 above)
Comment (GarrettBeatty) [reply .github/copilot-instructions.md:32]: good question. So this is assuming that copilot has some already pretrained info on .NET and AWS SDK coding practices. I'm also assuming that during PR review it has context of the files and it can infer existing best practices as well. 

you are right though that having a dedicated coding standards doc would help more though. let me know your thoughts.

File: .github/copilot-instructions.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+ - Review code for bugs, flaws in logic, and potential edge cases.
+ - Check for correct error handling and input validation.
+ - Ensure code is maintainable, readable, and follows best practices.
+ - Look for opportunities to simplify or optimize code where appropriate.
+
+## Automated Checks
+- Copilot should flag PRs that:
+ 	- Introduce breaking changes without a major version bump.
+ 	- Lack a dev config file or changelog update.
```
Comment (boblodgett) [inline .github/copilot-instructions.md]: changelog updates come through the devconfig files so it may not understand this is where it needs to see it. Again stressing it would be useful to build out the devconfig section above just a tad more.

File: .github/copilot-instructions.md (hunk 4 above)
Comment (GarrettBeatty) [reply .github/copilot-instructions.md:36]: So this is just based off of existing comments it gave e.g https://github.com/aws/aws-sdk-net/pull/3949#pullrequestreview-3098146159 it seems to know what it is and i believe it can infer (e.g. if file in the generator folder is modified then it refers to the generator)

File: .github/copilot-instructions.md (hunk 2 above)
Comment (boblodgett) [reply .github/copilot-instructions.md:16]: Yes, that should work. Otherwise it will have no clue what it is.

Comment (GarrettBeatty) [reply .github/copilot-instructions.md:16]: updated it to add a link to the contributing.md file (the link should resolve once the pr is merged to main)

Comment (GarrettBeatty) [reply .github/copilot-instructions.md:16]: Im not 100% sure if it can access links/other files but maybe we can try this way. Otherwise i will have to duplicate the contents in this file

File: .github/copilot-instructions.md (hunk 5 above)
Comment (GarrettBeatty) [reply .github/copilot-instructions.md:41]: updated

**Review by boblodgett — APPROVED** [review]
